packages/flutter_tools/lib/src/_flutter_manifest.dart

```diff
@@ -0,0 +1,460 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:pub_semver/pub_semver.dart';
+import 'package:yaml/yaml.dart';
+
+import 'base/file_system.dart';
+import 'base/user_messages.dart';
+import 'base/utils.dart';
+import 'globals.dart';
+
+/// The type of the Flutter project created using `flutter create -t <type>`
+enum ProjectType {
+  /// A Flutter project is considered an app when it doesn't a `module:`
+  /// or `plugin:` descriptor.
+  ///
+  /// Such a project can be created using `flutter create -t app`.
+  app,
+
+  /// A Flutter project is considered a module when it has a `module:`
+  /// descriptor. A Flutter module project supports integration into an
+  /// existing host app, and has managed platform host code.
+  ///
+  /// Such a project can be created using `flutter create -t module`.
+  module,
+
+  /// A Flutter project is considered a plugin when it has a `plugin:`
+  /// descriptor. A Flutter plugin project wraps custom Android and/or
+  /// iOS code in a Dart interface for consumption by other Flutter app
+  /// projects.
+  ///
+  /// Such a project can be created using `flutter create -t plugin`.
+  plugin,
+}
+
+/// A wrapper around the `flutter:` section in the `pubspec.yaml` file.
+final class FlutterManifest {
+  FlutterManifest({
+    this.appName = '',
+    this.appVersion = '',
+    this.buildName = '',
+    this.buildNumber = '',
+    this.usesMaterialDesign = false,
+    this.projectType = ProjectType.app,
+    this.assets = const <Uri>[],
+    this.fonts = const <Font>[],
+  });
+
+  /// The string value of the top-level `name` property in the `pubspec.yaml` file.
+  final String appName;
+
+  /// The version String from the `pubspec.yaml` file.
+  /// Can be null if it isn't set or has a wrong format.
+  final String appVersion;
```

Comment:
Even though this is documented as allowing null, it is defaulted to an empty String. This will required developers to explicitly pass null, which I think would be more surprising than a value that remains "unset" if not provided